### DISCUSSION TITLE
[Feature request] Close output window on successful compile

### DISCUSSION THREAD

**Discussion by javl** [glenn2223/vscode-live-sass-compiler#354] [Ideas]
I've been using this plugin for a while now and it is great. Especially with the `show output window on` set to `warning` so the window only appears when something goes wrong, alerting you of the problem.

But I was wondering: would it be possible to implement the opposite? An option that allows you to HIDE the output when compilation succeeds? 

Currently when you make a typo the output will appear to tell you something is wrong. You fix the error and then have to manually close the output again with this fiddly little `x` icon in the corner (unless I'm missing a shortcut).

It would be great if you could fix and save the code, triggering a new compile which, if succeeded, will close the output again.

**Comment by glenn2223** [comment]
So, it is definitely possible. There's a `hide` method that can be called.

There are a couple of details I want to hash out - to make sure I'm getting the expectations 100% correct - so, for now, I'll change this from an "issue" to a "discussion".

---

Ref (for me later): https://code.visualstudio.com/api/references/vscode-api#1783

**Comment by glenn2223** [comment]
So my initial thoughts/questions

### Integration

I can't think of a way this can be integrated into the current solution. If you can, let me know.

Because of this, a new setting must be introduced.  
A possible name could be `closeOutputWindowOnSuccess`.
- In your opinion, does this give an accurate representation of the functionality of the setting?
- If you were to see this as a new user, does the name make sense?
- If no to either, do you have a suggestion?

### Tieing it in

I just wanted to write out how I think the new functionality will work with the output window solution

It should work 2-fold (depending on how `showOutputWindowOn` is set):
- `Warning` and `Error`  
   This will serve the original request. That is, close the output window that was opened after an error was thrown
- `None`  
  Though there's no "normal" output. Critical extension issues are reported in certain scenarios. As it's usually a settings problem, upon saving a SASS file will be treated as that above (if the issue is resolved)
- `Trace`, `Debug` and `Information`  
  This one may be a problem. The perfect scenario is that the output window opens and closes in a way that isn't too jarring. However, smaller projects may have quick compilations that does have this effect - or the calls are so quick that it doesn't even bother trying to open it. **_Testing required_**

---

Feel free to "reply" directly to this comment. If you have your own questions/concerns, please open a new comment I can reply to. Thanks

**Reply by javl** [reply]
Your suggested behavior for `warning` and `error` seem exactly what I'd expect:

1. No output window open, editing along
2. Save the file and compiling fails because of an error in the code
3. The output window opens, showing you what's wrong
4. You fix the code and save again
5. The plugin recompiles, sees it was successful and closes the output window. If unsuccessful, the window will stay open to show the (new) error / warning.

Not sure about `None`, `Trace`, `Debug` and `Information`, as I don't use those modes and I'm unfamiliar with their exact usage, but I can imagine you don't really want to close the output window when you're using one of those. You use `debug` to see as much info as possible, so closing the window won't make much sense?

Maybe `closeOutputWindowOnSuccess` should simply be ignored for any mode beside `warning` and `error`.

**Reply by glenn2223** [reply]
You didn't say but I assume you're happy with clarity of the naming and what not

> Maybe `closeOutputWindowOnSuccess` should simply be ignored for any mode beside `warning` and `error`.

My issue with effectivly disabling a setting for certain states (especially default ones) could cause users to come charging in opening issues.

With my suggestion it could be treated as an alternate means of having `showOutputWindowOn` set to `Warning`.

I had a look at the code and I could implement it this way: if true, it holds off opening the output alltogether. Then, when everything has compiled, it opens the output at that point (depending on setings). If false, it continues as normal - in that it considers whether to open the ouput immediately (again, depending on setings)

**Reply by javl** [reply]
> You didn't say but I assume you're happy with clarity of the naming and what not

Sorry; yeah I think they make sense. You could go for `hideOutputWindowOn`, with `hide` being the opposite of `showOutputWindowOn` (instead of `closeOutputWindowOn` which would suggest `openOutputWindowOn`), but thats just nitpicking ;)

Your fear of getting loads of issues when it doesn't work as expected seems valid, but I still don't think closing the output window when using `debug` etc. would make a lot of sense though. But again, I don't use those modes so I might be mistaken. 

Isn't this something that could be solved with a small description in the settings, as you need to check the settings for the other modes anyway? 

> `closeOutputWindowOnSuccess`: if true, the output window will close after compiling successfully. Only used when `showOutputWindowOn` is set to `warning` or `error`.

**Reply by glenn2223** [reply]
> Sorry; yeah I think they make sense. You could go for `hideOutputWindowOn`, with `hide` being the opposite of `showOutputWindowOn` (instead of `closeOutputWindowOn` which would suggest `openOutputWindowOn`), but thats just nitpicking ;)

Good shout on the name. `hideOutputWindowOnSuccess` does fit in better.

> Your fear of getting loads of issues when it doesn't work as expected seems valid, but I still don't think closing the output window when using `debug` etc. would make a lot of sense though. But again, I don't use those modes so I might be mistaken.

Fair point with the `debug` - and with that `trace` too. I'll think on this; maybe for `information`, `warning` and `error`

---

I'm happy with everything we've discussed here so I'll create an issue for the feature request

Thanks for your help on ironing out all the details
